flink-java/src/main/java/org/apache/flink/api/java/functions/SemanticPropUtil.java

```diff
@@ -40,38 +45,108 @@
+	private final static String REGEX_FIELD_OR_FORWARD = "(" + REGEX_NESTED_FIELDS + "|" + REGEX_FORWARD + ")";
+	private final static String REGEX_ANNOTATION = "((" + REGEX_FIELD_OR_FORWARD + ";)*(" + REGEX_FIELD_OR_FORWARD + ");?)";
+
+	private static final Pattern PATTERN_WILDCARD = Pattern.compile(REGEX_WILDCARD);
 	private static final Pattern PATTERN_FORWARD = Pattern.compile(REGEX_FORWARD);
 	private static final Pattern PATTERN_ANNOTATION = Pattern.compile(REGEX_ANNOTATION);
 	private static final Pattern PATTERN_LIST = Pattern.compile(REGEX_LIST);
+	private static final Pattern PATTERN_FIELD = Pattern.compile(REGEX_NESTED_FIELDS);
 
-	private static final Pattern PATTERN_DIGIT = Pattern.compile("\\d+");
-
-	public static SingleInputSemanticProperties createProjectionPropertiesSingle(int[] fields) {
+	public static SingleInputSemanticProperties createProjectionPropertiesSingle(int[] fields, CompositeType<?> inType)
+	{
 		SingleInputSemanticProperties ssp = new SingleInputSemanticProperties();
-		for (int i = 0; i < fields.length; i++) {
-			ssp.addForwardedField(fields[i], i);
+
+		int[] sourceOffsets = new int[inType.getArity()];
+		sourceOffsets[0] = 0;
+		for(int i=1; i<inType.getArity(); i++) {
+			sourceOffsets[i] = inType.getTypeAt(i-1).getTotalFields() + sourceOffsets[i-1];
 		}
+
+		int targetOffset = 0;
+		for(int i=0; i<fields.length; i++) {
+			int sourceOffset = sourceOffsets[fields[i]];
+			int numFieldsToCopy = inType.getTypeAt(fields[i]).getTotalFields();
+
+			for(int j=0; j<numFieldsToCopy; j++) {
+				ssp.addForwardedField(sourceOffset+j, targetOffset+j);
+			}
+			targetOffset += numFieldsToCopy;
+		}
+
 		return ssp;
 	}
 
-	public static DualInputSemanticProperties createProjectionPropertiesDual(int[] fields, boolean[] isFromFirst) {
+	public static DualInputSemanticProperties createProjectionPropertiesDual(int[] fields, boolean[] isFromFirst,
+																				TypeInformation<?> inType1, TypeInformation<?> inType2)
+	{
 		DualInputSemanticProperties dsp = new DualInputSemanticProperties();
 
-		for (int i = 0; i < fields.length; i++) {
-			if (isFromFirst[i]) {
-				dsp.addForwardedField1(fields[i], i);
+		int[] sourceOffsets1;
+		if(inType1 instanceof TupleTypeInfo<?>) {
+			sourceOffsets1 = new int[inType1.getArity()];
+			sourceOffsets1[0] = 0;
+			for(int i=1; i<inType1.getArity(); i++) {
+				sourceOffsets1[i] = ((TupleTypeInfo<?>)inType1).getTypeAt(i-1).getTotalFields() + sourceOffsets1[i-1];
+			}
+		} else {
+			sourceOffsets1 = new int[] {0};
+		}
+
+		int[] sourceOffsets2;
+		if(inType2 instanceof TupleTypeInfo<?>) {
+			sourceOffsets2 = new int[inType2.getArity()];
+			sourceOffsets2[0] = 0;
+			for(int i=1; i<inType2.getArity(); i++) {
+				sourceOffsets2[i] = ((TupleTypeInfo<?>)inType2).getTypeAt(i-1).getTotalFields() + sourceOffsets2[i-1];
+			}
+		} else {
+			sourceOffsets2 = new int[] {0};
+		}
```

Comment:
Line 97 - 117 look pretty copy-pasted. I think a method that gets a TupleTypeInfo and returns the sourceOffset would be nicer here.